### PR TITLE
[TGUI] Convert ATM to TGUI

### PR DESCRIPTION
**Pull request by variableundefined**
## What Does This PR Do

Convert over ATM to TGUI interfaces. Also one minor typo fix in the Account Uplink terminal, and added dollar sign to the transaction log. 

**There is one functional change in this PR:**
- All PIN number are standardized to 6 digit instead of 4 digit, and uses the full range of 6 digit number instead of only up to 111111, which is plain weird.

Transaction log code wholeheartedly copy-pasted from the Accounts Uplink.

## Images of changes

**New:** 
image
image
image
image
image
**Old:** 
image
image
image
image
image

## Changelog
:cl:
tweak: Updated ATM interface to TGUI.
tweak: All PIN number are now six digit.
/:cl:

### CODE REVIEW

**Comment by ppi13** [comment]
Will this require a test merge?

**Comment by AffectedArc07** [comment]
> Will this require a test merge?

No, why would it

File: code/modules/economy/ATM.dm [hunk 1]
```diff
@@ -115,20 +115,20 @@ log transactions
...
-	ui_interact(user)
+	tgui_interact(user)
 
 /obj/machinery/atm/attack_ghost(mob/user)
-	ui_interact(user)
+	tgui_interact(user)
 
-/obj/machinery/atm/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+/obj/machinery/atm/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
 	user.set_machine(src)
```
Comment (AffectedArc07) [inline code/modules/economy/ATM.dm:124]:
```suggestion

```
TGUI removes the requirement for `set_machine()` (and so does nano, this is just super old code)

File: code/modules/economy/ATM.dm [hunk 2]
```diff
@@ -115,20 +115,20 @@ log transactions
...
 	if (!ui)
-		ui = new(user, src, ui_key, "atm.tmpl", name, 550, 650)
+		ui = new(user, src, ui_key, "ATM", name, 550, 650)
 		ui.open()
 
-/obj/machinery/atm/ui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.default_state)
-	var/data[0]
+/obj/machinery/atm/tgui_data(mob/user)
+	var/list/data = list()
 	data["src"] = UID()
```
Comment (AffectedArc07) [inline code/modules/economy/ATM.dm:132]:
```suggestion

```
Not needed anymore

File: tgui/packages/tgui/interfaces/ATM.js [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+const IntroductionAndCard = (props, context) => {
+  const { act, data } = useBackend(context);
+  const {
+    machine_id,
+    held_card_name,
+  } = data;
+  return (
+    <Section title="Nanotrasen Automatic Teller Machine" >
+      <Box>For all your monetary need!
+      </Box>
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/ATM.js:75]:
```suggestion
      <Box>
        For all your monetary need!
      </Box>
```

File: tgui/packages/tgui/interfaces/AccountsUplinkTerminal.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import { createSearch, decodeHtmlEntities } from 'common/string';
 import { Fragment } from 'inferno';
 import { useBackend, useLocalState } from "../backend";
-import { Box, Button, Flex, Icon, Input, LabeledList, NumberInput, Section, Table, Tabs } from '../components';
+import { Box, Divider, Button, Flex, Icon, Input, LabeledList, NumberInput, Section, Table, Tabs } from '../components';
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/AccountsUplinkTerminal.js:4]: Any reason for this import?

File: tgui/packages/tgui/interfaces/AccountsUplinkTerminal.js [hunk 5]
```diff
@@ -193,7 +193,7 @@ const DetailedAccountInfo = (properties, context) => {
           <LabeledList.Item label="Account Holder">
             {owner_name}
           </LabeledList.Item>
-          <LabeledList.Item label="Account Balnce">
+          <LabeledList.Item label="Account Balance">
```
Comment (AffectedArc07) [inline tgui/packages/tgui/interfaces/AccountsUplinkTerminal.js:196]: Technically this is a Bal~~a~~nce change

**Review by AffectedArc07 — CHANGES_REQUESTED** [review]

File: code/modules/economy/ATM.dm [hunk 6]
```diff
@@ -155,138 +153,144 @@ log transactions
...
+									T.target_name = failed_account.owner_name
+									T.purpose = "Unauthorised login attempt"
+									T.source_terminal = machine_id
+									T.date = GLOB.current_date_string
+									T.time = station_time_timestamp()
+									failed_account.transaction_log.Add(T)
 							else
-								to_chat(usr, "[bicon(src)]<span class='warning'>Incorrect pin/account combination entered.</span>")
-								number_incorrect_tries = 0
+								to_chat(usr, "[bicon(src)]<span class='warning'>Incorrect pin/account combination entered, [max_pin_attempts - number_incorrect_tries] attempts remaining.</span>")
```
Comment (dearmochi) [inline code/modules/economy/ATM.dm:210]:
```suggestion
								to_chat(usr, "[bicon(src)]<span class='warning'>Incorrect pin/account combination entered, [max_pin_attempts - number_incorrect_tries] attempt\s remaining.</span>")
```

File: code/modules/economy/ATM.dm [hunk 7]
```diff
@@ -155,138 +153,144 @@ log transactions
...
+					var/tried_pin = text2num(params["account_pin"])
+
+					authenticated_account = attempt_account_access(tried_account_num, tried_pin, held_card && held_card.associated_account_number == tried_account_num ? 2 : 1)
+					if(!authenticated_account)
+						number_incorrect_tries++
+						if(previous_account_number == tried_account_num)
+							if(number_incorrect_tries > max_pin_attempts)
+								//lock down the atm
+								ticks_left_locked_down = 30
+								playsound(src, 'sound/machines/buzz-two.ogg', 50, 1)
```
Comment (dearmochi) [inline code/modules/economy/ATM.dm:197]:
```suggestion
								playsound(src, 'sound/machines/buzz-two.ogg', 50, TRUE)
```

File: code/modules/economy/ATM.dm [hunk 8]
```diff
@@ -155,138 +153,144 @@ log transactions
...
+									T.source_terminal = machine_id
+									T.date = GLOB.current_date_string
+									T.time = station_time_timestamp()
+									failed_account.transaction_log.Add(T)
 							else
-								to_chat(usr, "[bicon(src)]<span class='warning'>Incorrect pin/account combination entered.</span>")
-								number_incorrect_tries = 0
+								to_chat(usr, "[bicon(src)]<span class='warning'>Incorrect pin/account combination entered, [max_pin_attempts - number_incorrect_tries] attempts remaining.</span>")
+								previous_account_number = tried_account_num
+								playsound(src, 'sound/machines/buzz-sigh.ogg', 50, 1)
```
Comment (dearmochi) [inline code/modules/economy/ATM.dm:212]:
```suggestion
								playsound(src, 'sound/machines/buzz-sigh.ogg', 50, TRUE)
```

File: code/modules/economy/ATM.dm [hunk 9]
```diff
@@ -155,138 +153,144 @@ log transactions
...
-
-				playsound(loc, pick('sound/items/polaroid1.ogg', 'sound/items/polaroid2.ogg'), 50, 1)
-			if("insert_card")
-				if(held_card)
-					held_card.forceMove(loc)
-					authenticated_account = null
-					if(ishuman(usr) && !usr.get_active_hand())
-						usr.put_in_hands(held_card)
-					held_card = null
+						playsound(src, 'sound/machines/twobeep.ogg', 50, 1)
```
Comment (dearmochi) [inline code/modules/economy/ATM.dm:217]:
```suggestion
						playsound(src, 'sound/machines/twobeep.ogg', 50, TRUE)
```

File: code/modules/economy/ATM.dm [hunk 10]
```diff
@@ -155,138 +153,144 @@ log transactions
...
+						ticks_left_timeout = 120
```
Comment (dearmochi) [inline code/modules/economy/ATM.dm:218]: Make that an option or define?

File: code/modules/economy/ATM.dm [hunk 11]
```diff
@@ -155,138 +153,144 @@ log transactions
...
+						to_chat(usr, "[bicon(src)]<span class='notice'>Access granted. Welcome user '[authenticated_account.owner_name].'</span>")
+					previous_account_number = tried_account_num
+
+		if("withdrawal")
+			var/amount = max(text2num(params["funds_amount"]),0)
+			if(amount <= 0)
+				alert("That is not a valid amount.")
+			else if(authenticated_account && amount > 0)
+				if(amount <= authenticated_account.money)
+					playsound(src, 'sound/machines/chime.ogg', 50, 1)
```
Comment (dearmochi) [inline code/modules/economy/ATM.dm:238]:
```suggestion
					playsound(src, 'sound/machines/chime.ogg', 50, TRUE)
```

File: code/modules/economy/ATM.dm [hunk 12]
```diff
@@ -155,138 +153,144 @@ log transactions
...
-			if("logout")
+					to_chat(usr, "[bicon(src)]<span class='warning'>You don't have enough funds to do that!</span>")
+
+		if("balance_statement")
+			if(authenticated_account)
+				if(world.timeofday < lastprint + PRINT_DELAY)
+					to_chat(usr, "<span class='notice'>The [name] flashes an error on its display.</span>")
+					return
+				lastprint = world.timeofday
+				playsound(loc, 'sound/goonstation/machines/printer_thermal.ogg', 50, 1)
```
Comment (dearmochi) [inline code/modules/economy/ATM.dm:255]:
```suggestion
				playsound(loc, 'sound/goonstation/machines/printer_thermal.ogg', 50, TRUE)
```

File: code/modules/economy/ATM.dm [hunk 13]
```diff
@@ -155,138 +153,144 @@ log transactions
...
+				//stamp the paper
+				var/image/stampoverlay = image('icons/obj/bureaucracy.dmi')
+				stampoverlay.icon_state = "paper_stamp-cent"
+				if(!R.stamped)
+					R.stamped = new()
+				R.stamped += /obj/item/stamp
+				R.overlays += stampoverlay
+				R.stamps += "<HR><i>This paper has been stamped by the Automatic Teller Machine.</i>"
+
+			playsound(loc, pick('sound/items/polaroid1.ogg', 'sound/items/polaroid2.ogg'), 50, 1)
```
Comment (dearmochi) [inline code/modules/economy/ATM.dm:274]:
```suggestion
			playsound(loc, pick('sound/items/polaroid1.ogg', 'sound/items/polaroid2.ogg'), 50, TRUE)
```

**Review by dearmochi — COMMENTED** [review]
A few things

File: code/modules/economy/ATM.dm [hunk 14]
```diff
@@ -155,138 +154,150 @@ log transactions
...
+				if(istype(I, /obj/item/card/id))
+					usr.drop_item()
+					I.forceMove(src)
+					held_card = I
+
+		if("logout")
+			authenticated_account = null
 
-	SSnanoui.update_uis(src)
+	. = TRUE
```
Comment (AffectedArc07) [inline code/modules/economy/ATM.dm:300]: Having `. = TRUE` and `return 1` is redundant here. Pick one or the other

**Review by AffectedArc07 — APPROVED** [review]

**Pull request merged**